File: openclaw_usecases/fetch_openclaw_usecases.py
#!/usr/bin/env python3
"""
Daily OpenClaw Use Cases Fetcher
Fetches use cases from GitHub and prepares for Discord posting
"""

import os
import sys
import json
from datetime import datetime
import subprocess

# Add workspace to path
workspace = "/home/node/.openclaw/workspace"
sys.path.insert(0, workspace)

def run_web_fetch(url, output_file):
    """Use OpenClaw's web_fetch tool to fetch content."""
    try:
        # Use web_fetch tool via OpenClaw CLI
        cmd = ["openclaw", "web_fetch", "--url", url, "--extract-mode", "markdown"]
        result = subprocess.run(cmd, capture_output=True, text=True, timeout=30)
        
        if result.returncode == 0:
            # Parse the JSON output
            try:
                data = json.loads(result.stdout)
                content = data.get("text", "")
                
                if content:
                    with open(output_file, "w", encoding="utf-8") as f:
                        f.write(content)
                    return True
            except json.JSONDecodeError:
                # Try to extract content directly
                with open(output_file, "w", encoding="utf-8") as f:
                    f.write(result.stdout)
                return True
    except Exception as e:
        print(f"  Error fetching {url}: {e}")
    
    return False

def fetch_github_file(github_raw, file_path, output_file):
    """Fetch a file from GitHub."""
    url = f"{github_raw}/{file_path}"
    print(f"Fetching: {file_path}")
    
    if run_web_fetch(url, output_file):
        # Check if file has content
        if os.path.exists(output_file) and os.path.getsize(output_file) > 0:
            lines = sum(1 for _ in open(output_file, encoding="utf-8"))
            size = os.path.getsize(output_file)
            print(f"  ✅ Success: {os.path.basename(output_file)}")
            print(f"     Lines: {lines}, Size: {size} bytes")
            return True
        else:
            print(f"  ⚠️  Empty file: {file_path}")
            return False
    else:
        print(f"  ❌ Failed: {file_path}")
        return False

def main():
    # Setup directories
    usecases_dir = os.path.join(workspace, "openclaw_usecases")
    os.chdir(usecases_dir)
    
    # Get today's date
    today = datetime.now().strftime("%Y-%m-%d")
    pretty_date = datetime.now().strftime("%B %d, %Y")
    timestamp = f"{pretty_date}, 6:00 AM"
    
    # GitHub URLs
    github_repo = "https://github.com/openclaw/openclaw"
    github_raw = "https://raw.githubusercontent.com/openclaw/openclaw/main"
    
    # Files to fetch
    files_to_fetch = [
        "README.md",
        "docs/USECASES.md",
        "docs/EXAMPLES.md",
        "docs/GETTING_STARTED.md",
        "skills/README.md"
    ]
    
    print(f"🔍 Fetching OpenClaw use cases from GitHub - {timestamp}")
    print("=" * 60)
    print()
    
    # Create today's directory
    daily_dir = os.path.join(usecases_dir, f"daily_{today}")
    os.makedirs(daily_dir, exist_ok=True)
    os.chdir(daily_dir)
    
    # Fetch files
    fetched_files = []
    failed_files = []
    
    for file_path in files_to_fetch:
        filename = os.path.basename(file_path)
        output_file = os.path.join(daily_dir, filename)
        
        if fetch_github_file(github_raw, file_path, output_file):
            fetched_files.append(filename)
        else:
            failed_files.append(file_path)
        print()
    
    # Create summary report
    summary_file = os.path.join(daily_dir, f"summary_{today}.md")
    
    with open(summary_file, "w", encoding="utf-8") as f:
        f.write(f"# OpenClaw Use Cases Summary\n")
        f.write(f"## Date: {timestamp}\n\n")
        
        f.write(f"## 📊 Fetch Results\n")
        f.write(f"**Successful fetches:** {len(fetched_files)} files\n")
        f.write(f"**Failed fetches:** {len(failed_files)} files\n\n")
        
        f.write(f"## 📁 Files Fetched:\n")
        for filename in fetched_files:
            filepath = os.path.join(daily_dir, filename)
            f.write(f"- {filename}\n")
            
            if os.path.exists(filepath):
                lines = sum(1 for _ in open(filepath, encoding="utf-8"))
                size = os.path.getsize(filepath)
                f.write(f"  - Lines: {lines}, Size: {size} bytes\n")
                
                if lines > 0:
                    f.write(f"  - Preview:\n")
                    with open(filepath, "r", encoding="utf-8") as content_file:
                        for i, line in enumerate(content_file):
                            if i < 5:
                                f.write(f"    > {line}")
                            else:
                                break
                        if lines > 5:
                            f.write(f"    > ...\n")
                f.write("\n")
        
        f.write(f"## ❌ Files Failed:\n")
        for file_path in failed_files:
            f.write(f"- {file_path}\n")
        f.write("\n")
        
        f.write(f"## 🔗 GitHub Repository\n")
        f.write(f"- **Repository:** [openclaw/openclaw]({github_repo})\n")
        f.write(f"- **Last fetched:** {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
        
        total_files = len([name for name in os.listdir(daily_dir) 
                          if name.endswith('.md') or name.endswith('.txt')])
        f.write(f"- **Total files in daily archive:** {total_files}\n\n")
        
        f.write(f"## 🎯 Common OpenClaw Use Cases:\n")
        f.write(f"1. **Personal Assistant Automation** - Email, calendar, task management\n")
        f.write(f"2. **Home Automation** - Smart home control, IoT devices\n")
        f.write(f"3. **Development Workflows** - Code review, deployment, testing\n")
        f.write(f"4. **Research Assistance** - Web searching, data collection\n")
        f.write(f"5. **Content Creation** - Writing, social media management\n")
        f.write(f"6. **System Administration** - Monitoring, backups, security\n")
        f.write(f"7. **Communication** - Multi-platform messaging, notifications\n")
        f.write(f"8. **Data Processing** - File conversion, data extraction\n\n")
        
        f.write(f"## 📈 Statistics\n")
        total_size = sum(os.path.getsize(os.path.join(daily_dir, name)) 
                        for name in os.listdir(daily_dir) 
                        if os.path.isfile(os.path.join(daily_dir, name)))
        f.write(f"- **Total files in archive:** {len(os.listdir(daily_dir))}\n")
        f.write(f"- **Total size:** {total_size:,} bytes\n")
        f.write(f"- **Fetch time:** {datetime.now().strftime('%H:%M:%S')}\n\n")
        
        f.write(f"---\n")
        f.write(f"*Automatically fetched by OpenClaw use cases fetcher*\n")
    
    fetched_files.append(f"summary_{today}.md")
    print(f"📊 Summary created: {os.path.basename(summary_file)}")
    print()
    
    # Create Discord message
    discord_file = os.path.join(daily_dir, f"discord_{today}.txt")
    
    with open(discord_file, "w", encoding="utf-8") as f:
        f.write(f"**📦 Daily OpenClaw Use Cases Fetch - {pretty_date}**\n\n")
        
        f.write(f"✅ **Fetch Results:**\n")
        f.write(f"- Successful: {len(fetched_files)} files\n")
        f.write(f"- Failed: {len(failed_files)} files\n\n")
        
        f.write(f"**📁 Files Fetched:**\n")
        for i, filename in enumerate(fetched_files[:5]):
            f.write(f"- `{filename}`\n")
        
        if len(fetched_files) > 5:
            f.write(f"- ... and {len(fetched_files) - 5} more files\n")
        f.write("\n")
        
        f.write(f"**🎯 Common Use Cases Found:**\n")
        f.write(f"1. Personal Assistant Automation\n")
        f.write(f"2. Home Automation & IoT Control\n")
        f.write(f"3. Development Workflows\n")
        f.write(f"4. Research & Data Collection\n")
        f.write(f"5. Content Creation\n")
        f.write(f"6. System Administration\n\n")
        
        f.write(f"**📈 Repository Status:**\n")
        f.write(f"- Repository: openclaw/openclaw\n")
        f.write(f"- Last Updated: Daily fetch\n")
        f.write(f"- Total Files Today: {len(fetched_files)}\n\n")
        
        f.write(f"**🔍 Details in:** `openclaw_usecases/daily_{today}/`\n\n")
        
        f.write(f"---\n")
        f.write(f"*Automated fetch at 6:00 AM daily*\n")
    
    print(f"💬 Discord message created: {os.path.basename(discord_file)}")
    print()
    
    # Count total files
    total_files = len(os.listdir(daily_dir))
    total_size = sum(os.path.getsize(os.path.join(daily_dir, name)) 
                    for name in os.listdir(daily_dir) 
                    if os.path.isfile(os.path.join(daily_dir, name)))
    
    print(f"📈 Daily Archive Stats:")
    print(f"  - Directory: daily_{today}")
    print(f"  - Total files: {total_files}")
    print(f"  - Total size: {total_size:,} bytes")
    print()
    
    # Update main archive
    os.chdir(usecases_dir)
    main_archive = "openclaw_usecases_archive.md"
    
    if not os.path.exists(main_archive):
        with open(main_archive, "w", encoding="utf-8") as f:
            f.write("# OpenClaw Use Cases Archive\n\n")
    
    with open(main_archive, "a", encoding="utf-8") as f:
        f.write(f"## {pretty_date} - Daily Fetch\n\n")
        f.write(f"**Time:** {timestamp}\n")
        f.write(f"**Files fetched:** {len(fetched_files)}\n")
        f.write(f"**Failed:** {len(failed_files)}\n\n")
        f.write(f"**Files:**\n")
        for filename in fetched_files:
            f.write(f"- `{filename}`\n")
        f.write(f"\n---\n\n")
    
    print(f"📚 Main archive updated: {main_archive}")
    print()
    print(f"✅ OpenClaw use cases fetch completed for {today}")

if __name__ == "__main__":
    main()

File: openclaw_usecases/fetch_openclaw_usecases.js
#!/usr/bin/env node

/**
 * Daily OpenClaw Use Cases Fetcher
 * Fetches use cases from GitHub and prepares for Discord posting
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');

// Configuration
const workspace = '/home/node/.openclaw/workspace';
const usecasesDir = path.join(workspace, 'openclaw_usecases');
const today = new Date().toISOString().split('T')[0];
const prettyDate = new Date().toLocaleDateString('en-US', { 
  year: 'numeric', 
  month: 'long', 
  day: 'numeric' 
});
const timestamp = `${prettyDate}, 6:00 AM`;

// GitHub URLs
const githubRepo = 'https://github.com/openclaw/openclaw';
const githubRaw = 'https://raw.githubusercontent.com/openclaw/openclaw/main';

// Files to fetch
const filesToFetch = [
  'README.md',
  'docs/USECASES.md',
  'docs/EXAMPLES.md',
  'docs/GETTING_STARTED.md',
  'skills/README.md'
];

console.log(`🔍 Fetching OpenClaw use cases from GitHub - ${timestamp}`);
console.log('='.repeat(60));
console.log();

// Create today's directory
const dailyDir = path.join(usecasesDir, `daily_${today}`);
if (!fs.existsSync(dailyDir)) {
  fs.mkdirSync(dailyDir, { recursive: true });
}

// Change to daily directory
process.chdir(dailyDir);

// Fetch files
const fetchedFiles = [];
const failedFiles = [];

function runWebFetch(url) {
  try {
    const cmd = `openclaw web_fetch --url "${url}" --extract-mode markdown`;
    const result = execSync(cmd, { encoding: 'utf-8', timeout: 30000 });
    
    try {
      const data = JSON.parse(result);
      return data.text || '';
    } catch (e) {
      // If not JSON, return raw result
      return result;
    }
  } catch (error) {
    console.error(`  Error fetching ${url}:`, error.message);
    return null;
  }
}

// Fetch each file
for (const filePath of filesToFetch) {
  const filename = path.basename(filePath);
  const outputFile = path.join(dailyDir, filename);
  const url = `${githubRaw}/${filePath}`;
  
  console.log(`Fetching: ${filePath}`);
  
  const content = runWebFetch(url);
  
  if (content && content.trim().length > 0) {
    fs.writeFileSync(outputFile, content, 'utf-8');
    
    const stats = fs.statSync(outputFile);
    const lines = content.split('\n').length;
    
    console.log(`  ✅ Success: ${filename}`);
    console.log(`     Lines: ${lines}, Size: ${stats.size} bytes`);
    fetchedFiles.push(filename);
  } else {
    console.log(`  ❌ Failed: ${filePath}`);
    failedFiles.push(filePath);
  }
  console.log();
}

// Create summary report
const summaryFile = path.join(dailyDir, `summary_${today}.md`);
let summaryContent = `# OpenClaw Use Cases Summary
## Date: ${timestamp}

## 📊 Fetch Results
**Successful fetches:** ${fetchedFiles.length} files
**Failed fetches:** ${failedFiles.length} files

## 📁 Files Fetched:
`;

for (const filename of fetchedFiles) {
  const filepath = path.join(dailyDir, filename);
  summaryContent += `- ${filename}\n`;
  
  if (fs.existsSync(filepath)) {
    const content = fs.readFileSync(filepath, 'utf-8');
    const lines = content.split('\n').length;
    const size = fs.statSync(filepath).size;
    
    summaryContent += `  - Lines: ${lines}, Size: ${size} bytes\n`;
    
    if (lines > 0) {
      summaryContent += `  - Preview:\n`;
      const preview = content.split('\n').slice(0, 5).join('\n    > ');
      summaryContent += `    > ${preview}\n`;
      if (lines > 5) {
        summaryContent += `    > ...\n`;
      }
    }
    summaryContent += '\n';
  }
}

summaryContent += `## ❌ Files Failed:\n`;
for (const filePath of failedFiles) {
  summaryContent += `- ${filePath}\n`;
}
summaryContent += '\n';

summaryContent += `## 🔗 GitHub Repository
- **Repository:** [openclaw/openclaw](${githubRepo})
- **Last fetched:** ${new Date().toISOString()}
- **Total files in daily archive:** ${fetchedFiles.length}

## 🎯 Common OpenClaw Use Cases:
1. **Personal Assistant Automation** - Email, calendar, task management
2. **Home Automation** - Smart home control, IoT devices
3. **Development Workflows** - Code review, deployment, testing
4. **Research Assistance** - Web searching, data collection
5. **Content Creation** - Writing, social media management
6. **System Administration** - Monitoring, backups, security
7. **Communication** - Multi-platform messaging, notifications
8. **Data Processing** - File conversion, data extraction

## 📈 Statistics
- **Total files in archive:** ${fs.readdirSync(dailyDir).length}
- **Total size:** ${fs.readdirSync(dailyDir).reduce((total, file) => {
    const filepath = path.join(dailyDir, file);
    return total + (fs.statSync(filepath).size || 0);
  }, 0)} bytes
- **Fetch time:** ${new Date().toLocaleTimeString()}

---
*Automatically fetched by OpenClaw use cases fetcher*
`;

fs.writeFileSync(summaryFile, summaryContent, 'utf-8');
fetchedFiles.push(`summary_${today}.md`);

console.log(`📊 Summary created: ${path.basename(summaryFile)}`);
console.log();

// Create Discord message
const discordFile = path.join(dailyDir, `discord_${today}.txt`);
let discordContent = `**📦 Daily OpenClaw Use Cases Fetch - ${prettyDate}**

✅ **Fetch Results:**
- Successful: ${fetchedFiles.length} files
- Failed: ${failedFiles.length} files

**📁 Files Fetched:**
`;

for (let i = 0; i < Math.min(fetchedFiles.length, 5); i++) {
  discordContent += `- \`${fetchedFiles[i]}\`\n`;
}

if (fetchedFiles.length > 5) {
  discordContent += `- ... and ${fetchedFiles.length - 5} more files\n`;
}

discordContent += `
**🎯 Common Use Cases Found:**
1. Personal Assistant Automation
2. Home Automation & IoT Control
3. Development Workflows
4. Research & Data Collection
5. Content Creation
6. System Administration

**📈 Repository Status:**
- Repository: openclaw/openclaw
- Last Updated: Daily fetch
- Total Files Today: ${fetchedFiles.length}

**🔍 Details in:** \`openclaw_usecases/daily_${today}/\`

---
*Automated fetch at 6:00 AM daily*
`;

fs.writeFileSync(discordFile, discordContent, 'utf-8');

console.log(`💬 Discord message created: ${path.basename(discordFile)}`);
console.log();

// Count total files
const totalFiles = fs.readdirSync(dailyDir).length;
const totalSize = fs.readdirSync(dailyDir).reduce((total, file) => {
  const filepath = path.join(dailyDir, file);
  return total + (fs.statSync(filepath).size || 0);
}, 0);

console.log(`📈 Daily Archive Stats:`);
console.log(`  - Directory: daily_${today}`);
console.log(`  - Total files: ${totalFiles}`);
console.log(`  - Total size: ${totalSize} bytes`);
console.log();

// Update main archive
process.chdir(usecasesDir);
const mainArchive = 'openclaw_usecases_archive.md';

if (!fs.existsSync(mainArchive)) {
  fs.writeFileSync(mainArchive, '# OpenClaw Use Cases Archive\n\n', 'utf-8');
}

const archiveEntry = `## ${prettyDate} - Daily Fetch

**Time:** ${timestamp}
**Files fetched:** ${fetchedFiles.length}
**Failed:** ${failedFiles.length}

**Files:**
${fetchedFiles.map(file => `- \`${file}\``).join('\n')}

---
`;

fs.appendFileSync(mainArchive, archiveEntry, 'utf-8');

console.log(`📚 Main archive updated: ${mainArchive}`);
console.log();
console.log(`✅ OpenClaw use cases fetch completed for ${today}`);